opm/upscaling/RelPermUtils.hpp

```diff
@@ -0,0 +1,161 @@
+
+#ifndef OPM_UPSCALING_RELPERM_UTILS_HPP
+#define OPM_UPSCALING_RELPERM_UTILS_HPP
+
+#include <opm/core/utility/MonotCubicInterpolator.hpp>
+#include <opm/upscaling/SinglePhaseUpscaler.hpp>
+#include <array>
+#include <tuple>
+#include <vector>
+
+namespace Opm {
+  //! \brief Get value from tensor
+  //! \param[in] K The tensor to extract the value from
+  //! \param[in] voigt_idx The voigt index for value to extract (0..8)
+  //! \return The requested value
+  double getVoigtValue(const SinglePhaseUpscaler::permtensor_t& K, int voigt_idx);
+
+  //! \brief Set value in tensor
+  //! \param[out] K The tensor to set value in
+  //! \param[in] voigt_idx The voigt index for value to set (0..8)
+  //! \param[in] val Value to set in tensor
+  void setVoigtValue(SinglePhaseUpscaler::permtensor_t& K, int voigt_idx, double val);
+
+  //! \brief Helper class for relperm upscaling applications.
+  class RelPermUpscaleHelper {
+    public:
+      bool isMaster;                               //!< Whether this is the master MPI node or not.
+      bool doEclipseCheck;                         //!< Whether to check that input relperm curves include relperm at critical saturation points.
+      bool anisotropic_input;                      //!< Whether input eclipse file has diagonal anisotrophy.
+      int points;                                  //!< Number of saturation points to upscale for.
+      int tensorElementCount;                      //!< Number of independent elements in resulting tensor.
+      bool upscaleBothPhases;                      //!< Whether to upscale both phases.
+      std::vector<double> WaterSaturation;         //!< Re-upscaled water saturation for the computed pressure points.
+      SinglePhaseUpscaler::permtensor_t permTensor; //!< Tensor of upscaled results.
+      std::vector<int> satnums;                    //!< Cell satnums.
+      std::vector<MonotCubicInterpolator> InvJfunctions; //!< Inverse of the loaded J-functions.
+      //! \brief Relperm-curves for each (component->phase->stone type)
+      std::array<std::array<std::vector<MonotCubicInterpolator>,2>,3> Krfunctions;
+      SinglePhaseUpscaler::BoundaryConditionType boundaryCondition; //!< Boundary conditions to use.
+      std::vector<MonotCubicInterpolator> SwPcfunctions; //!< Holds Sw(Pc) for each rocktype.
+      std::string saturationstring; //!< Fluid system type.
+      size_t tesselatedCells; //!< Number of "active" cells (Sintef interpretation of "active")
+      double volume; //!< Total volume.
+      double poreVolume; //!< Total pore volume.
+      std::vector<double> pressurePoints; //!< Vector of capillary pressure points between Swor and Swir.
+
+      //! \brief Default constructor.
+      //! \param[in] mpi_rank Rank of this process (for parallel simulations).
+      //! \param[in] options Options structure.
+      //! \details Uses the following options: fluids
+      RelPermUpscaleHelper(int mpi_rank, std::map<std::string,std::string>& options_);
```

Comment:
Why is the map passed as a non-const reference (also: data member)? Is it to use operator[]?